src/exo/LoopIR_scheduling.py

```diff
@@ -2564,11 +2564,147 @@ def map_e(self, e):
+            if e.op == '/' or e.op == '%':
+                return True
+            else:
+                lhs = self.has_div_mod(e.lhs)
+                rhs = self.has_div_mod(e.rhs)
+                return lhs or rhs
+        else:
+            assert False, "bad case"
+
+    # Call this when e is one indexing expression
+    # e should be an indexing expression
+    def index_start(self, e):
+        assert isinstance(e, LoopIR.expr)
+        # Div and mod need more subtle handling. Don't normalize for now.
+        if self.has_div_mod(e):
+            return e
+
+        # Make a map of symbols and coefficients
+        n_map = self.normalize_e(e)
+
+        # Write back to LoopIR.expr
+        def get_loopir(key, value):
+            vconst = LoopIR.Const(value, T.int, e.srcinfo)
+            if key == self.C:
+                return vconst
+            else:
+                readkey = LoopIR.Read(key, [], e.type, e.srcinfo)
+                return LoopIR.BinOp('*', vconst, readkey, e.type, e.srcinfo)
+        
+        delete_zero = { key: n_map[key] for key in n_map if n_map[key] != 0 }
+        new_e = LoopIR.Const(0, T.int, e.srcinfo)
+        for key, val in delete_zero.items():
+            if val > 0:
+                # add
+                new_e = LoopIR.BinOp('+', new_e, get_loopir(key, val), e.type, e.srcinfo)
+            else:
+                # sub
+                new_e = LoopIR.BinOp('-', new_e, get_loopir(key, val), e.type, e.srcinfo)
+
+        return new_e
+
+
+    def map_e(self, e):
+        if isinstance(e, LoopIR.Read):
+            if len(e.idx) > 0:
+                # This is a Tensor read
+                new_idx = [ self.index_start(ei) for ei in e.idx ]
+                return LoopIR.Read(e.name, new_idx, e.type, e.srcinfo)
+
+        return super().map_e(e)
+
+    def map_w_access(self, w):
+        if isinstance(w, LoopIR.Interval):
+            return LoopIR.Interval(self.index_start(w.lo),
+                                   self.index_start(w.hi), w.srcinfo)
+        else:
+            return LoopIR.Point(self.index_start(w.pt), w.srcinfo)
+
+    def map_s(self, s):
+        if isinstance(s, (LoopIR.Assign, LoopIR.Reduce)):
+            new_idx = [ self.index_start(e) for e in s.idx ]
+            return [type(s)(s.name, s.type, s.cast, new_idx, self.map_e(s.rhs),
+                            s.eff, s.srcinfo)]
+        elif isinstance(s, (LoopIR.ForAll, LoopIR.Seq)):
+            new_hi = self.index_start(s.hi)
+            new_body = self.map_stmts(s.body)
+            return [type(s)(s.iter, new_hi, new_body, s.eff, s.srcinfo)]
+        elif isinstance(s, LoopIR.Alloc):
+            if isinstance(s.type, T.Tensor):
+                new_hi = [ self.index_start(e) for e in s.type.hi ]
+                new_typ = T.Tensor(new_hi, False, s.type.basetype())
+                return [LoopIR.Alloc(s.name, new_typ, s.mem, s.eff, s.srcinfo)]
+
+        return super().map_s(s)
```

Comment:
Could you eliminate all of these different `map_*` cases by simply checking at the top of `map_e` whether or not an expression is of `index` or `size` type (i.e. is `indexable`) and then returning the result of `index_start` if it is?

Reply:
Yes, thanks!